src/gui/mzroll/downloadmanager.cpp

```diff
@@ -0,0 +1,86 @@
+#include "downloadmanager.h"
+#include <QNetworkAccessManager>
+#include <QNetworkRequest>
+#include  <QNetworkReply>
+#include <QDebug>
+#include "pollyintegration.h"
+
+DownloadManager::DownloadManager():
+    _manager(new QNetworkAccessManager),
+    _reply(nullptr),
+    err(false)
+{
+}
+
+void DownloadManager::setRequest(const QString &url, void *requester, bool async)
+{
+    _url = url;
+    //reset the data for every url
+    _data = "";
+    download(async);
+}
+
+void DownloadManager::download(bool async)
+{
+    QNetworkRequest request(_url);
+    _reply = _manager->get(request);
+
+    if(!async) {
+        // synchronous request
+        QEventLoop loop;
+        connect(_reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
+        connect(_reply, QOverload<QNetworkReply::NetworkError>::of(&QNetworkReply::error),
+                &loop, &QEventLoop::quit);
+        loop.exec();
+        if(_reply->error() == QNetworkReply::NoError) {
+            _data += _reply->readAll();
+            err = false;
+            qDebug() << "sync request complete";
+        } else {
+            qDebug() << "error code : " << _reply->error();
+            qDebug() << "error : " << _reply->errorString();
+            err = true;
+
+        }
+    } else {
+        connect(_reply, &QNetworkReply::readyRead, this, &DownloadManager::dataAvailable);
+        connect(_reply, &QNetworkReply::finished, this, &DownloadManager::finished);
+        connect(_reply, QOverload<QNetworkReply::NetworkError>::of(&QNetworkReply::error),
+                this, &DownloadManager::error);
+    }
+}
+
+QByteArray DownloadManager::getData()
+{
+    return _data;
+}
+
+
+void DownloadManager::dataAvailable()
+{
+    qDebug() << "data available";
+    while(_reply->bytesAvailable()) {
+        _data += _reply->readLine();
+    }
+}
+
+void DownloadManager::finished()
+{
+    if(err) {
+        qDebug() << _reply->errorString();
+         emit failed();
+    } else {
+        // read any data that might be in buffer
+        _data += _reply->readAll();
+        qDebug() << "success";
+        emit downloaded();
+    }
+}
+
+
+void DownloadManager::error(QNetworkReply::NetworkError error)
+{
+    qDebug() << "error code :" << error;
+    qDebug() << "error string: "<< _reply->errorString();
+    err = true;
```

Comment:
Don't you think storing the error string might be helpful, instead of just printing them? It could help us to inform the user on GUI or CLI about what the precise error was. The string can be empty if there was no error. We could add a checker (e.g. `errorOccurred`) to check whether the request terminated in error state, and we won't need the boolean `err` value.

Reply:
We should ideally

Reply:
Yes. But since we are short on time, we can make this change when its actually needed. Resolving.